galsim/wfirst/wfirst_wcs.py

```diff
@@ -0,0 +1,478 @@
+
+        # Go from the tangent plane position of the SCA center, to the actual celestial coordinate,
+        # using `world_pos` as the center point of the tangent plane projection.  This celestial
+        # coordinate for the SCA center is `crval`, which goes into the WCS as CRVAL1, CRVAL2.
+        u = -sca_xc_tp_f * cos_pa - sca_yc_tp_f * sin_pa
+        v = -sca_xc_tp_f * sin_pa + sca_yc_tp_f * cos_pa
+        crval = world_pos.deproject(galsim.PositionD(u/galsim.arcsec, v/galsim.arcsec),
+                                    projection='gnomonic')
+        crval1 = crval.ra
+        crval2 = crval.dec
+        header['CRVAL1'] = (crval1 / galsim.degrees, "first axis value at reference pixel")
+        header['CRVAL2'] = (crval2 / galsim.degrees, "second axis value at reference pixel")
+
+        # Compute the position angle of the local pixel Y axis.
+        # This requires projecting local North onto the detector axes.
+        # Start by adding any SCA-unique rotation relative to FPA axes:
+        sca_tp_rot = pa_fpa + sca_rot[i_sca]*galsim.degrees
+
+        # Set up a point that defines the vector going through the SCA center (crval) pointing North.
+        # Use a distance in declination that is halfway from the center to the edge of the SCA
+        # (local_dist). This seems a reasonable compromise in making the step big enough to avoid
+        # round-off errors becoming too large and avoiding non-linearities in the trigonometric
+        # functions.
+        local_dist = float(galsim.wfirst.n_pix) * pix_scale / 4
+        if crval2 + local_dist > 90.*galsim.degrees:
+            pos2 = galsim.CelestialCoord(crval1, crval2-local_dist)
+        else:
+            pos2 = galsim.CelestialCoord(crval1, crval2+local_dist)
+
+        # Convert to tangent plane X-Y positions (outputs are in arcsec and have opposite
+        # handedness, so must be careful of units and flip sign of x coordinate).  We only care
+        # about X-distance and Y-distance between pos2 and crval (position of SCA center), which
+        # tells us how the line that points north is oriented on the detector.
+        tp2 = world_pos.project(pos2, 'gnomonic')
+        if pos2.dec > crval2:
+            dxtp = -tp2.x + u/galsim.arcsec
+            dytp = tp2.y - v/galsim.arcsec
+        else:
+            dxtp = -u/galsim.arcsec+tp2.x
+            dytp = v/galsim.arcsec-tp2.y
+        # A few things to note about the next line.
+        # First, we appear to be putting the Delta(x) value in the place where the "y" coordinate
+        # should go for the np.arctan2, and Delta(y) value in the place of the "x" coordinate.  This
+        # is because of how our tangent plane (x, y) positions are defined: we define the rotation
+        # angle with respect to the -Y axis (not X).  The minus sign in front also comes from this
+        # convention (the fact that it's the -Y axis, not +Y).
+        pa_sca = sca_tp_rot - np.arctan2(dxtp, dytp)*galsim.radians
```

Comment:
OK, Now I see what this is better.  Here is the code I would use to compute it without having to worry about passing over due north:
```
# Go some reasonable distance from crval in the +y direction.  Say, 1 degree.
plusy = world_pos.deproject(galsim.PositionD(u/galsim.arcsec, v/galsim.arcsec + 3600),
                               projection='gnomonic')
# Find the angle between this point, crval and due north
north = galsim.CelestialCoord(0.*galsim.degrees, 90.*galsim.degrees)
pa_sca = sca_tp_rot - crval.angleBetween(plusy, north)
```
I checked that this gives the same answer as your procedure to ~ 1.e-8 level accuracy.  And it strikes me as a bit cleaner.